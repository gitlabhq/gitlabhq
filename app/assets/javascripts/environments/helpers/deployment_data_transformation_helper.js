import { getIdFromGraphQLId } from '~/graphql_shared/utils';

/**
 * This function transforms Commit object coming from GraphQL to object compatible with app/assets/javascripts/vue_shared/components/commit.vue author object
 * @param {Object} Commit
 * @returns {Object}
 */
export const getAuthorFromCommit = (commit) => {
  if (commit.author) {
    return {
      username: commit.author.name,
      path: commit.author.webUrl,
      avatar_url: commit.author.avatarUrl,
    };
  }
  return {
    username: commit.authorName,
    path: `mailto:${commit.authorEmail}`,
    avatar_url: commit.authorGravatar,
  };
};

/**
 * This function transforms deploymentNode object coming from GraphQL to object compatible with app/assets/javascripts/vue_shared/components/commit.vue
 * @param {Object} deploymentNode
 * @returns {Object}
 */
export const getCommitFromDeploymentNode = (deploymentNode) => {
  if (!deploymentNode.commit) {
    throw new Error("deploymentNode argument doesn't have 'commit' field", deploymentNode);
  }
  return {
    title: deploymentNode.commit.message,
    commitUrl: deploymentNode.commit.webUrl,
    shortSha: deploymentNode.commit.shortId,
    tag: deploymentNode.tag,
    commitRef: {
      name: deploymentNode.ref,
    },
    author: getAuthorFromCommit(deploymentNode.commit),
  };
};

export const convertJobToDeploymentAction = (job) => {
  return {
    name: job.name,
    playable: job.playable,
    scheduledAt: job.scheduledAt,
    playPath: `${job.webPath}/play`,
  };
};

export const getActionsFromDeploymentNode = (deploymentNode, lastDeploymentName) => {
  if (!deploymentNode || !lastDeploymentName) {
    return [];
  }

  return (
    deploymentNode.job?.deploymentPipeline?.jobs?.nodes
      ?.filter((deployment) => deployment.name !== lastDeploymentName)
      .map(convertJobToDeploymentAction) || []
  );
};

export const getRollbackActionFromDeploymentNode = (deploymentNode, environment) => {
  const { job, id } = deploymentNode;

  if (!job) {
    return null;
  }
  const isLastDeployment = id === environment.lastDeployment?.id;
  const { webPath } = job;
  return {
    id,
    name: environment.name,
    lastDeployment: {
      commit: deploymentNode.commit,
      isLast: isLastDeployment,
    },
    retryUrl: `${webPath}/retry`,
  };
};

const getDeploymentApprovalFromDeploymentNode = (deploymentNode, environment) => {
  if (!environment.protectedEnvironments || environment.protectedEnvironments.nodes.length === 0) {
    return {
      isApprovalActionAvailable: false,
    };
  }

  const protectedEnvironmentInfo = environment.protectedEnvironments.nodes[0];

  const hasApprovalRules = protectedEnvironmentInfo.approvalRules.nodes?.length > 0;
  const hasRequiredApprovals = protectedEnvironmentInfo.requiredApprovalCount > 0;

  const isApprovalActionAvailable = hasRequiredApprovals || hasApprovalRules;
  const requiredMultipleApprovalRulesApprovals =
    protectedEnvironmentInfo.approvalRules.nodes.reduce((requiredApprovals, rule) => {
      return requiredApprovals + rule.requiredApprovals;
    }, 0);

  const requiredApprovalCount = hasRequiredApprovals
    ? protectedEnvironmentInfo.requiredApprovalCount
    : requiredMultipleApprovalRulesApprovals;

  return {
    isApprovalActionAvailable,
    deploymentIid: deploymentNode.iid,
    environment: {
      name: environment.name,
      tier: environment.tier,
      requiredApprovalCount,
    },
  };
};

/**
 * This function transforms deploymentNode object coming from GraphQL to object compatible with app/assets/javascripts/environments/environment_details/page.vue table
 * @param {Object} deploymentNode
 * @returns {Object}
 */
export const convertToDeploymentTableRow = (deploymentNode, environment) => {
  const { lastDeployment } = environment;
  const commit = getCommitFromDeploymentNode(deploymentNode);
  const { job } = deploymentNode;
  const pipeline = job?.downstreamPipeline ? job?.downstreamPipeline : job?.deploymentPipeline;

  return {
    status: deploymentNode.status.toLowerCase(),
    id: deploymentNode.iid,
    triggerer: deploymentNode.triggerer,
    commit,
    job: job && {
      webPath: job.webPath,
      label: `${job.name} (#${getIdFromGraphQLId(job.id)})`,
      pipeline: pipeline?.path && {
        path: pipeline?.path,
        label: `#${getIdFromGraphQLId(pipeline.id)}`,
      },
    },
    created: deploymentNode.createdAt || '',
    finished: deploymentNode.finishedAt || '',
    actions: getActionsFromDeploymentNode(deploymentNode, lastDeployment?.job?.name),
    rollback: getRollbackActionFromDeploymentNode(deploymentNode, environment),
    deploymentApproval: getDeploymentApprovalFromDeploymentNode(deploymentNode, environment),
    webPath: deploymentNode.webPath || '',
  };
};
