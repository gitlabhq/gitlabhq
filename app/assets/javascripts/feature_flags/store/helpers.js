import { ROLLOUT_STRATEGY_GITLAB_USER_LIST, NEW_VERSION_FLAG } from '../constants';

const mapStrategyScopesToRails = (scopes) =>
  scopes.length === 0
    ? [{ environment_scope: '*' }]
    : scopes.map((s) => ({
        id: s.id,
        _destroy: s.shouldBeDestroyed,
        environment_scope: s.environmentScope,
      }));

const mapStrategyScopesToView = (scopes) =>
  scopes.map((s) => ({
    id: s.id,
    // eslint-disable-next-line no-underscore-dangle
    shouldBeDestroyed: Boolean(s._destroy),
    environmentScope: s.environment_scope,
  }));

const mapStrategiesParametersToViewModel = (params) => {
  if (params.userIds) {
    return { ...params, userIds: params.userIds.split(',').join(', ') };
  }
  return params;
};

export const mapStrategiesToViewModel = (strategiesFromRails) =>
  (strategiesFromRails || []).map((s) => ({
    id: s.id,
    name: s.name,
    parameters: mapStrategiesParametersToViewModel(s.parameters),
    userList: s.user_list,
    // eslint-disable-next-line no-underscore-dangle
    shouldBeDestroyed: Boolean(s._destroy),
    scopes: mapStrategyScopesToView(s.scopes),
  }));

const mapStrategiesParametersToRails = (params) => {
  if (params.userIds) {
    return { ...params, userIds: params.userIds.replace(/\s*,\s*/g, ',') };
  }
  return params;
};

const mapStrategyToRails = (strategy) => {
  const mappedStrategy = {
    id: strategy.id,
    name: strategy.name,
    _destroy: strategy.shouldBeDestroyed,
    scopes_attributes: mapStrategyScopesToRails(strategy.scopes || []),
    parameters: mapStrategiesParametersToRails(strategy.parameters),
  };

  if (strategy.name === ROLLOUT_STRATEGY_GITLAB_USER_LIST) {
    mappedStrategy.user_list_id = strategy.userList?.id;
  }
  return mappedStrategy;
};

export const mapStrategiesToRails = (params) => ({
  operations_feature_flag: {
    name: params.name,
    description: params.description,
    active: params.active,
    strategies_attributes: (params.strategies || []).map(mapStrategyToRails),
    version: NEW_VERSION_FLAG,
  },
});
