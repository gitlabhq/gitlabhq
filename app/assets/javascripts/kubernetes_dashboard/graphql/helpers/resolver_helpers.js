import {
  CoreV1Api,
  Configuration,
  WatchApi,
  EVENT_DATA,
  EVENT_TIMEOUT,
  EVENT_ERROR,
} from '@gitlab/cluster-client';
import { getWatchManager } from '~/environments/services/websocket_connection_service';
import { connectionStatus } from '~/environments/graphql/resolvers/kubernetes/constants';
import { updateConnectionStatus } from '~/environments/graphql/resolvers/kubernetes/k8s_connection_status';
import { s__ } from '~/locale';

export const handleClusterError = async (err) => {
  if (!err.response) {
    throw err;
  }

  const contentType = err.response.headers.get('Content-Type');

  if (contentType !== 'application/json') {
    throw new Error(
      s__(
        'KubernetesDashboard|There was a problem fetching cluster information. Refresh the page and try again.',
      ),
    );
  }

  const errorData = await err.response.json();
  throw errorData;
};

export const buildWatchPath = ({ resource, api = 'api/v1', namespace = '' }) => {
  return namespace ? `/${api}/namespaces/${namespace}/${resource}` : `/${api}/${resource}`;
};

export const mapWorkloadItem = (item) => {
  const status = item.status || {};
  const spec = item.spec || {};
  const metadata = {
    ...item.metadata,
    annotations: item.metadata?.annotations || {},
    labels: item.metadata?.labels || {},
  };
  return { status, spec, metadata, __typename: 'LocalWorkloadItem' };
};

export const mapEventItem = ({
  lastTimestamp = '',
  eventTime = '',
  message,
  reason,
  source,
  type,
}) => ({ lastTimestamp, eventTime, message, reason, source, type });

export const subscribeToSocket = async ({ watchId, watchParams, cacheParams, config }) => {
  const { updateQueryCache, updateConnectionStatusFn } = cacheParams;

  try {
    const watcherConnection = getWatchManager(config);

    const watcher = await watcherConnection.initConnection({
      message: { watchId, watchParams },
    });

    const handleConnectionStatus = (status) => {
      if (updateConnectionStatusFn) {
        updateConnectionStatusFn(status);
      }
    };

    watcher.on(EVENT_DATA, watchId, (data) => {
      updateQueryCache(data);
      handleConnectionStatus(connectionStatus.connected);
    });

    watcher.on(EVENT_ERROR, watchId, () => {
      handleConnectionStatus(connectionStatus.disconnected);
    });
  } catch (err) {
    throw new Error(s__('KubernetesDashboard|Failed to establish WebSocket connection'));
  }
};

export const watchWorkloadItems = async ({
  client,
  query,
  configuration,
  namespace,
  watchPath,
  queryField,
  watchParams,
}) => {
  const config = new Configuration(configuration);
  const watcherApi = new WatchApi(config);

  const updateStatus = (status) =>
    updateConnectionStatus(client, {
      configuration,
      namespace,
      resourceType: queryField,
      status,
    });

  const updateQueryCache = (data) => {
    const result = data.map(mapWorkloadItem);
    client.writeQuery({
      query,
      variables: { configuration, namespace },
      data: { [queryField]: result },
    });
  };

  const watchFunction = async () => {
    try {
      const watcher = await watcherApi.subscribeToStream(watchPath, { watch: true });

      watcher.on(EVENT_DATA, (data) => {
        updateQueryCache(data);
        updateStatus(connectionStatus.connected);
      });

      watcher.on(EVENT_TIMEOUT, () => {
        updateStatus(connectionStatus.disconnected);
      });

      watcher.on(EVENT_ERROR, () => {
        updateStatus(connectionStatus.disconnected);
      });
    } catch (err) {
      await handleClusterError(err);
    }
  };

  updateStatus(connectionStatus.connecting);

  const watchId = namespace ? `${queryField}-n-${namespace}` : `${queryField}-all-namespaces`;
  const cacheParams = {
    updateQueryCache,
    updateConnectionStatusFn: updateStatus,
  };

  try {
    await subscribeToSocket({ watchId, watchParams, cacheParams, config });
  } catch {
    await watchFunction();
  }
};

export const getK8sPods = ({
  client,
  query,
  configuration,
  namespace = '',
  mapFn = mapWorkloadItem,
  queryField = 'k8sPods',
}) => {
  const config = new Configuration(configuration);

  const coreV1Api = new CoreV1Api(config);
  const podsApi = namespace
    ? coreV1Api.listCoreV1NamespacedPod({ namespace })
    : coreV1Api.listCoreV1PodForAllNamespaces();

  return podsApi
    .then((res) => {
      const watchPath = buildWatchPath({ resource: 'pods', namespace });
      const watchParams = {
        version: 'v1',
        resource: 'pods',
        namespace,
      };

      watchWorkloadItems({
        client,
        query,
        configuration,
        namespace,
        watchPath,
        queryField,
        watchParams,
      });

      const data = res?.items || [];
      return data.map(mapFn);
    })
    .catch(async (err) => {
      try {
        await handleClusterError(err);
      } catch (error) {
        throw new Error(error.message);
      }
    });
};
