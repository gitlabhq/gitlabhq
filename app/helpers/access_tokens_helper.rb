# frozen_string_literal: true

module AccessTokensHelper
  include AccountsHelper
  include ApplicationHelper

  def scope_description(prefix)
    case prefix
    when :project_access_token
      [:doorkeeper, :project_access_token_scope_desc]
    when :group_access_token
      [:doorkeeper, :group_access_token_scope_desc]
    else
      [:doorkeeper, :scope_desc]
    end
  end

  def tokens_app_data
    {
      feed_token: {
        enabled: !Gitlab::CurrentSettings.disable_feed_token,
        token: current_user.feed_token,
        reset_path: reset_feed_token_profile_path
      },
      incoming_email_token: {
        enabled: incoming_email_token_enabled?,
        token: current_user.enabled_incoming_email_token,
        reset_path: reset_incoming_email_token_profile_path
      },
      static_object_token: {
        enabled: static_objects_external_storage_enabled?,
        token: current_user.enabled_static_object_token,
        reset_path: reset_static_object_token_profile_path
      }
    }.to_json
  end

  def filter_sort_scopes(scopes, sources)
    scopes.select { |scope| ::Gitlab::Auth::UI_SCOPES_ORDERED_BY_PERMISSION.include?(scope) }
    .sort_by { |scope| ::Gitlab::Auth::UI_SCOPES_ORDERED_BY_PERMISSION.index(scope) }
    .map do |value|
      {
        value: value,
        text: t(value, scope: sources)
      }
    end
  end

  def personal_access_token_data(token, user = current_user)
    sources = scope_description(:personal_access_token)
    scopes = ::Gitlab::Auth.available_scopes_for(user)
    {
      access_token: {
        **expires_at_field_data,
        available_scopes: filter_sort_scopes(scopes, sources).to_json,
        name: token[:name],
        description: token[:description],
        scopes: token[:scopes].to_json,
        create: user_settings_personal_access_tokens_url,
        revoke: expose_url(api_v4_personal_access_tokens_path),
        rotate: expose_url(api_v4_personal_access_tokens_path),
        show: "#{expose_url(api_v4_personal_access_tokens_path)}?user_id=:id"
      }
    }
  end

  def expires_at_field_data
    {
      min_date: 1.day.from_now.iso8601,
      max_date: max_date_allowed
    }
  end

  private

  def max_date_allowed
    return unless Gitlab::CurrentSettings.require_personal_access_token_expiry?

    ::PersonalAccessToken.max_expiration_lifetime_in_days.days.from_now.iso8601
  end
end

AccessTokensHelper.prepend_mod
