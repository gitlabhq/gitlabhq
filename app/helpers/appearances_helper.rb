# frozen_string_literal: true

module AppearancesHelper
  include MarkupHelper
  include Gitlab::Utils::StrongMemoize

  def appearance_pwa_icon_path_scaled(width)
    return unless Appearance::ALLOWED_PWA_ICON_SCALER_WIDTHS.include?(width)

    append_root_path((current_appearance&.pwa_icon_path_scaled(width) || "/-/pwa-icons/logo-#{width}.png"))
  end

  def appearance_maskable_logo
    append_root_path('/-/pwa-icons/maskable-logo.png')
  end

  def append_root_path(path)
    Gitlab::Utils.append_path(Gitlab.config.gitlab.relative_url_root, path)
  end

  def brand_title
    current_appearance&.title.presence || default_brand_title
  end

  def custom_sign_in_brand_title
    # no-op overridden in EE
  end

  def custom_sign_up_brand_title
    # no-op overridden in EE
  end

  def appearance_apple_touch_icon
    link_tags = favicon_link_tag('apple-touch-icon.png', rel: 'apple-touch-icon')

    return link_tags unless current_appearance&.pwa_icon.present?

    link_tags = favicon_link_tag(
      appearance_pwa_icon_path_scaled(Appearance::ALLOWED_PWA_ICON_SCALER_WIDTHS.first),
      rel: 'apple-touch-icon'
    )

    Appearance::ALLOWED_PWA_ICON_SCALER_WIDTHS.each do |width|
      link_tags += "\n"
      link_tags += favicon_link_tag(appearance_pwa_icon_path_scaled(width),
        sizes: "#{width}x#{width}", rel: 'apple-touch-icon')
    end

    link_tags
  end

  def appearance_pwa_name
    current_appearance&.pwa_name.presence || _('GitLab')
  end

  def appearance_pwa_short_name
    current_appearance&.pwa_short_name.presence || _('GitLab')
  end

  def appearance_pwa_description
    current_appearance&.pwa_description.presence ||
      _("The complete DevOps platform. " \
        "One application with endless possibilities. " \
        "Organizations rely on GitLabâ€™s source code management, " \
        "CI/CD, security, and more to deliver software rapidly.")
  end

  def default_brand_title
    # This resides in a separate method so that EE can easily redefine it.
    _('GitLab Community Edition')
  end

  def brand_image
    image_tag(brand_image_path, alt: brand_title, class: 'gl-invisible gl-h-10 js-portrait-logo-detection')
  end

  def brand_image_path
    return current_appearance.logo_path if current_appearance&.logo?

    image_path('logo.svg')
  end

  def custom_sign_in_description
    markdown_field(current_appearance, :description)
  end

  def brand_member_guidelines
    markdown_field(current_appearance, :member_guidelines)
  end

  def brand_new_project_guidelines
    markdown_field(current_appearance, :new_project_guidelines)
  end

  def brand_profile_image_guidelines
    markdown_field(current_appearance, :profile_image_guidelines)
  end

  def current_appearance
    strong_memoize(:current_appearance) do
      Appearance.current
    end
  end

  def brand_header_logo(options = {})
    add_gitlab_logo_text = options[:add_gitlab_logo_text] || false

    if current_appearance&.header_logo?
      image_tag current_appearance.header_logo_path, class: 'brand-header-logo', alt: ''
    elsif add_gitlab_logo_text
      render partial: 'shared/logo_with_text', formats: :svg
    else
      render partial: 'shared/logo', formats: :svg
    end
  end

  def header_message
    return unless current_appearance&.show_header?

    class_names = []
    class_names << 'with-performance-bar' if performance_bar_enabled?

    render_message(:header_message, class_names: class_names)
  end

  def footer_message
    return unless current_appearance&.show_footer?

    render_message(:footer_message)
  end

  private

  def render_message(field_sym, class_names: [], style: message_style)
    class_names << field_sym.to_s.dasherize

    content_tag :div, class: class_names, style: style do
      markdown_field(current_appearance, field_sym)
    end
  end

  def message_style
    style = []
    style << "background-color: #{current_appearance.message_background_color};"
    style << "color: #{current_appearance.message_font_color}"
    style.join
  end
end

AppearancesHelper.prepend_mod
