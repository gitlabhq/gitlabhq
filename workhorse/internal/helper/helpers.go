// Package helper provides utility functions for various tasks
package helper

import (
	"errors"
	"mime"
	"net/url"
	"os"
	"path/filepath"
	"strings"

	"gitlab.com/gitlab-org/gitlab/workhorse/internal/log"
)

// OpenFile opens a file at the specified path and returns its properties
func OpenFile(path string) (file *os.File, fi os.FileInfo, err error) {
	file, err = os.Open(filepath.Clean(path))
	if err != nil {
		return
	}

	defer func() {
		if err != nil {
			_ = file.Close()
		}
	}()

	fi, err = file.Stat()
	if err != nil {
		return
	}

	// The os.Open can also open directories
	if fi.IsDir() {
		err = &os.PathError{
			Op:   "open",
			Path: path,
			Err:  errors.New("path is directory"),
		}
		return
	}

	return
}

// URLMustParse parses the given string as a URL
func URLMustParse(s string) *url.URL {
	u, err := url.Parse(s)
	if err != nil {
		log.WithError(err).WithFields(log.Fields{"url": s}).Fatal("urlMustParse")
	}
	return u
}

// IsContentType checks if the actual content type matches the expected content type
func IsContentType(expected, actual string) bool {
	parsed, _, err := mime.ParseMediaType(actual)
	return err == nil && parsed == expected
}

// IsURL checks if the given string starts with http:// or https://
func IsURL(path string) bool {
	return strings.HasPrefix(path, "http://") || strings.HasPrefix(path, "https://")
}
