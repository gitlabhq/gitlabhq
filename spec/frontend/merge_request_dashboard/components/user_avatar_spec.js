import { shallowMountExtended } from 'helpers/vue_test_utils_helper';
import UserAvatar from '~/merge_request_dashboard/components/user_avatar.vue';

let wrapper;

function createComponent() {
  wrapper = shallowMountExtended(UserAvatar, {
    propsData: {
      user: {
        id: 'gid://gitlab/user/2',
        webUrl: '/root',
        name: 'Admin',
        avatarUrl: '/root',
        mergeRequestInteraction: {
          reviewState: 'REQUESTED_CHANGES',
        },
      },
    },
  });
}

describe('Merge request dashboard user avatar component', () => {
  const findReviewStateIcon = () => wrapper.findByTestId('review-state-icon');

  it('renders review state icon', () => {
    createComponent();

    expect(findReviewStateIcon().html()).toMatchSnapshot();
  });
});
