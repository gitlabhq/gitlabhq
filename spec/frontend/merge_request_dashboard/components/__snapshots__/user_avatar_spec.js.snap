// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Merge request dashboard user avatar component renders review state icon 1`] = `
<span
  class="-gl-bottom-2 -gl-left-2 gl-absolute gl-bg-status-danger gl-flex gl-h-5 gl-items-center gl-justify-center gl-p-1 gl-rounded-full gl-w-5"
  data-testid="review-state-icon"
>
  <gl-icon-stub
    class="gl-block gl-fill-status-danger"
    name="error"
    size="12"
    variant="current"
  />
</span>
`;
